ironfish/src/workerPool/tasks/decryptNotes.ts

```diff
@@ -0,0 +1,242 @@
+    const reader = bufio.read(buffer, true)
+
+    const hasCurrentNoteIndex = Boolean(reader.readU8())
+    const serializedNote = reader.readBytes(ENCRYPTED_NOTE_LENGTH)
+    const incomingViewKey = reader.readBytes(KEY_LENGTH).toString('hex')
+    const outgoingViewKey = reader.readBytes(KEY_LENGTH).toString('hex')
+    const spendingKey = reader.readBytes(KEY_LENGTH).toString('hex')
+
+    let currentNoteIndex = null
+    if (hasCurrentNoteIndex) {
+      currentNoteIndex = reader.readU32()
+    }
+
+    return new DecryptNotesRequest(
+      serializedNote,
+      incomingViewKey,
+      outgoingViewKey,
+      spendingKey,
+      currentNoteIndex,
+      jobId,
+    )
+  }
+
+  getSize(): number {
+    let size = 1 + ENCRYPTED_NOTE_LENGTH + KEY_LENGTH + KEY_LENGTH + KEY_LENGTH
+    if (this.currentNoteIndex) {
+      size += 4
+    }
+    return size
+  }
+}
+
+export class DecryptNotesResponse extends WorkerMessage {
+  readonly note: DecryptedNote | null
+
+  constructor(note: DecryptedNote | null, jobId: number) {
+    super(WorkerMessageType.DecryptNotes, jobId)
+    this.note = note
+  }
+
+  serialize(): Buffer {
+    const bw = bufio.write(this.getSize())
+
+    const hasDecryptedNote = Number(!!this.note)
+    bw.writeU8(hasDecryptedNote)
+
+    if (this.note) {
+      let flags = 0
+      flags |= Number(!!this.note.index) << 0
+      flags |= Number(!!this.note.nullifier) << 1
+      flags |= Number(this.note.forSpender)
+      bw.writeU8(flags)
+      bw.writeHash(this.note.merkleHash)
+      bw.writeBytes(this.note.serializedNote)
+
+      if (this.note.index) {
+        bw.writeU32(this.note.index)
+      }
+
+      if (this.note.nullifier) {
+        bw.writeHash(this.note.nullifier)
+      }
+    }
+
+    return bw.render()
+  }
+
+  static deserialize(jobId: number, buffer: Buffer): DecryptNotesResponse {
+    const reader = bufio.read(buffer)
+
+    const hasDecryptedNote = reader.readU8()
+    if (!hasDecryptedNote) {
+      return new DecryptNotesResponse(null, jobId)
+    }
+
+    const flags = reader.readU8()
+    const hasIndex = flags & (1 << 0)
+    const hasNullifier = flags & (1 << 1)
+    const forSpender = Boolean(flags & (1 << 2))
+    const merkleHash = reader.readHash()
+    const serializedNote = reader.readBytes(NOTE_LENGTH)
+
+    let index = null
+    if (hasIndex) {
+      index = reader.readU32()
+    }
+
+    let nullifier = null
+    if (hasNullifier) {
+      nullifier = reader.readHash()
+    }
+
+    return new DecryptNotesResponse(
+      {
+        forSpender,
+        index,
+        merkleHash,
+        nullifier,
+        serializedNote,
+      },
+      jobId,
+    )
+  }
+
+  getSize(): number {
+    let size = 1
+
+    if (this.note) {
+      size += 1 + 32 + NOTE_LENGTH
+
+      if (this.note.index) {
+        size += 4
+      }
+
+      if (this.note.nullifier) {
+        size += 32
+      }
+    }
+
+    return size
+  }
+}
+
+export class DecryptNotesTask extends WorkerTask {
```

Comment:
This is plural but only takes a single note

Reply:
Keeping as is since this now takes multiple payloads for decryption